src/util/psbt/responsibilities.rs

```diff
@@ -0,0 +1,147 @@
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Implementation of PSBT responsibilities according to BIP-174
+
+use std::marker::PhantomData;
+use std::error::Error;
+use std::fmt::{Formatter, Display};
+use secp256k1::{Message, Signature};
+use util::bip32::{Fingerprint, DerivationPath};
+use super::PartiallySignedTransaction;
+
+/// Trait for a generic PSBT responsibility (decorator+facade)
+pub trait Responsibility where Self: Sized {
+    /// Constructs responsibility decorator/facade from a given PSBT
+    fn from_psbt(psbt: PartiallySignedTransaction) -> Self;
+
+    /// Converts decorator/facade into the raw PSBT object
+    fn into_psbt(self) -> PartiallySignedTransaction;
+
+    /// Returns PSBT representation behind the decorator/facade
+    fn as_psbt(&self) -> &PartiallySignedTransaction;
+}
+
+macro_rules! impl_responsibility {
+    ($ident:ident) => {
+        impl Responsibility for $ident {
+            fn from_psbt(psbt: PartiallySignedTransaction) -> Self {
+                Self { psbt }
+            }
+
+            fn into_psbt(self) -> PartiallySignedTransaction {
+                self.psbt
+            }
+
+            fn as_psbt(&self) -> &PartiallySignedTransaction {
+                &self.psbt
+            }
+        }
+    };
+}
+
+/// PSBT signing verification error
+#[derive(Debug)]
+pub enum VerificationError {
+}
+impl Display for VerificationError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        Ok(())
+    }
+}
+impl Error for VerificationError {}
+
+/// PSBT signing error
+#[derive(Debug)]
+pub enum SigningError {
+    /// Can't sign because PSBT structure is invalid
+    VerificationError(VerificationError)
+}
+impl Display for SigningError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SigningError::VerificationError(err) => err.fmt(f)
+        }
+    }
+}
+impl Error for SigningError {}
+
+
+impl From<VerificationError> for SigningError {
+    fn from(err: VerificationError) -> Self {
+        SigningError::VerificationError(err)
+    }
+}
+
+/// Trait for PSBT signing responsibility
+pub trait SignPsbt: Responsibility {
+    /// Concrete verification error type
+    type VerificationError: Error;
+
+    /// Concrete signing error type
+    type SigningError: Error + From<Self::VerificationError>;
+
+    /// Verifies PSBT structure against BIp-174 rules
+    fn verify(&self) -> Result<(), Self::VerificationError> {
+        // TODO: Some default validation code
+        Ok(())
+    }
+
+    /// Signs all PSBT inputs
+    fn sign(self) -> Result<PartiallySignedTransaction, Self::SigningError>;
+}
+
+/// Default implementation for PSBT signing
+pub struct DefaultSigner<F, E, SigningInfo = Message>
+    where F: Fn(SigningInfo, Fingerprint, DerivationPath) -> Result<Option<Signature>, E>,
+          E: Error
+{
+    psbt: PartiallySignedTransaction,
```

Comment:
Why would you want the signer to own the PSBT?

Reply:
Responsibilities add state machine over PBST; meaning that PBST can be only is a particular responsibility state, not their mix.

Reply:
Does explicitly modeling the state machine give us any _big_ advantages? I'm all for enforcing invariants through the type system, but the PSBT process isn't all that complicated and at least in the signing stage the guarantees it can give are limited (multiple signing rounds).

If we really want to go this way of enforcing the correct life cycle of a PSBT through the type system I'd build it using newtypes (kinda like you did), but in the following way:

* Every state a PSBT can be is has a newtype with
  * a constructor that checks the invariants and returns a `Result`
  * a function to turn it back into a plain PSBT
  * e.g. a PSBT that can be signed is in the state `updated` and stays there until it's finalized
* Trivial state transitions that don't require external data (combiner, finalizer, extractor) can be modeled using member functions of the newtypes that represent the source state and transforms it into the target state, possibly returning an error
* Nontrivial state transitions (update, sign, maybe create?) that need external data should be modeled as traits that define the state transition as a function consuming the source state newtype and returning a result with the target state newtype
  * we may supply a default impl, e.g. an extended private key or any type implementing `Fn` with a certain signature is a signer or an updater based on bitcoincore-rpc (if that makes sense)
  * but we always leave the door open for other implementations and allow abstracting over them (e.g. this function needs an argument that can act as a PSBT signer, we don't care how it works inside)

This would add a lot of complexity, but _might_ be worth it. One annoying thing with this is that external implementors of these state transition trait's can't be allowed to directly create state newtypes (to guarantee the invariants). They'll have to go through the fallible constructors and probably `unwrap` if they know they did everything correctly.

Reply:
Hm, this is exactly I am trying to do, I do not see where we are different that much (agree with constructor that checks invariants). I just drafted a single state related to signing, but just to align on APIs, before implementing the rest.

Reply:
> EDIT: diagram of the state transitions as I imagine them

My intent was to implement them according to the spec BIP-174, where they are quite different

Reply:
> My intent was to implement them according to the spec BIP-174, where they are quite different

That might be where our disagreement stems from, how do you interpret BIP-174? I see a few points where my model doesn't cover all possible workflows (multiple/out of order updates), but if we take that into account will we still gain any safety from modeling a workflow? When you look at this example from the BIP:

![PSBT example workflow](https://raw.githubusercontent.com/bitcoin/bips/master/bip-0174/coinjoin-workflow.svg)

There's a giant state in which arbitrary updates/signatures can happen and there's no sense in modeling two independent states. You'd be left with the `created` and `finalized` state. But who cares about the finalized state anyway? All we want is the extracted BTC tx at that point. That's part of the reason why I'm doubtful that we should implement these states in the type system.